sklearn/cluster/som_.py

```diff
@@ -0,0 +1,248 @@
+"""
+ Self-organizing map
+"""
+
+# Authors: Sebastien Campion <sebastien.campion@inria.fr>
+#          naught101 <naught101@gmail.com>
+# License: BSD
+
+from __future__ import division
+import numpy as np
+from ..base import BaseEstimator
+from ..decomposition import PCA
+
+
+def _unserialise_coordinate(serial, spacing):
+    coord = []
+    for i in range(len(spacing)):
+        coord.append(int(np.floor(serial / spacing[i])))
+        serial = int(serial % spacing[i])
+
+    return(coord)
+
+
+def _serialise_coordinate(coord, spacing):
+    return(np.dot(coord, spacing))
+
+
+def _generate_adjacency_matrix(grid_dimensions):
+    """Generate an adjacency matrix of a rectangular grid.
+
+    grid_dimensions : tuple of integers
+    Specifies the dimensions of the grid for which to generate an adjacency
+    matrix.
+    """
+    n_centers = np.prod(grid_dimensions)
+
+    # initialise to 0 for self-distances, infinity non-connections
+    adjacency = np.empty((n_centers, n_centers), dtype=float)
+    adjacency.fill(np.inf)
+    np.fill_diagonal(adjacency, 0)
+
+    spacing = [int(np.prod(grid_dimensions[(k + 1):]))
+               for k in range(len(grid_dimensions))]
+
+    for i in range(n_centers):
+        coord = _unserialise_coordinate(i, spacing)
+        neighbours = []
+        for d in range(len(grid_dimensions)):
+            if coord[d] > 0:
+                down = coord[:]
+                down[d] -= 1
+                neighbours.append(down)
+            if coord[d] < (grid_dimensions[d] - 1):
+                up = coord[:]
+                up[d] += 1
+                neighbours.append(up)
+
+        for neighbour in neighbours:
+            serial = _serialise_coordinate(neighbour, spacing)
+            adjacency[i, serial] = 1
+
+    return(adjacency)
+
+
+def _get_minimum_distances(adjacency):
+    """Generate an all-pairs shortest path matrix from an adjacency matrix.
+
+    adjacency : symmetric square ndarray
+    Matrix with distances between pairs of graph nodes (usually 1s), or np.inf
+    for no connection.
+
+    Based on the `Floyd-Warshall algorithm
+    <https://en.wikipedia.org/wiki/Floyd-Warshall_algorithm>`_.
+    """
+    assert (len(adjacency.shape) == 2 and adjacency.shape[0] == adjacency.shape[1]), \
+        "adjacency isn't a square matrix"
+    n_nodes = adjacency.shape[0]
+    distance = adjacency.copy()
+    for k in range(n_nodes):
```

Comment:
This should use a scipy function methinks. Currently it is a n_nodes ** 3 loop in python, which seems bad.